agent/consul/rate/handler.go

```diff
@@ -166,11 +183,34 @@ func (h *Handler) Run(ctx context.Context) {
 // Allow returns an error if the given operation is not allowed to proceed
 // because of an exhausted rate-limit.
 func (h *Handler) Allow(op Operation) error {
-	// TODO(NET-1383): actually implement the rate limiting logic.
-	//
-	// Example:
-	//	if !h.limiter.Allow(globalWrite) {
-	//	}
+	for _, l := range h.limits(op) {
+		if l.mode == ModeDisabled {
+			continue
+		}
+
+		if h.limiter.Allow(l.ent) {
+			continue
+		}
+
+		// TODO: metrics.
+		// TODO: is this the correct log-level?
+		enforced := l.mode == ModeEnforcing
+		h.logger.Trace("RPC exceeded allowed rate limit",
+			"rpc", op.Name,
+			"source_addr", op.SourceAddr.String(),
+			"limit_type", l.desc,
+			"limit_enforced", enforced,
+		)
+
+		if enforced {
+			if h.delegate.IsLeader() {
```

Comment:
Because we register a delegate after the server creation, I think we should have a safe guard in Allow in case delegate is nil. Also, I think we should make delegate an atomic to avoid races.

Reply:
Good idea, `handler.Run(...)` now panics if `handler.RegisterDelegate(...)` hasn't been set.

I'm not sure I understand why `delegate` should be an atomic. `delegate.IsLeader()` is read-only and the underlying implementation ends up calling `rate.GetState(...)`.

```Go
func (r *raftState) getState() RaftState {
	stateAddr := (*uint32)(&r.state)
	return RaftState(atomic.LoadUint32(stateAddr))
}
```

Reply:
I think we are also missing a check here about the request being a write, a snippet from the RFC:

```
The only exception to this will be write requests rate-limited in the leader node 
by a global limit which will return a UNAVAILABLE response.
The reason for this is that if a write request hit a global limit in the leader node:
It’s not a retriable error, as only the leader is able to handle this request
The leader reached a critical limit
```

If that not possible right now, because of dependencies in other PRs, maybe we can add a TODO around it.

Reply:
Good catch! Fixed in https://github.com/hashicorp/consul/pull/15857/commits/2a936ab9202fde18b13341bf00970b693ac77b2a